spec/unit/typesafe/config/tokenizer_spec.rb

```diff
@@ -0,0 +1,696 @@
+require 'spec_helper'
+require 'hocon'
+require 'test_utils'
+require 'pp'
+
+
+describe Hocon::Impl::Tokenizer do
+  Tokens = Hocon::Impl::Tokens
+
+  shared_examples_for "token_matching" do
+    it "should match the tokenized string to the list of expected tokens" do
+      tokenized_from_string = TestUtils.tokenize_as_list(test_string)
+
+      # Add START and EOF tokens
+      wrapped_tokens = TestUtils.wrap_tokens(expected_tokens)
+
+      # puts "\n\n===Expected String: #{test_string}"
+      # PP.pp(tokenized_from_string)
+
+      # Compare the two lists of tokens
+      expect(tokenized_from_string).to eq(wrapped_tokens)
+    end
+  end
+
+  shared_examples_for "strings_with_problems" do
+    it "should find a problem when tokenizing" do
+      token_list = TestUtils.tokenize_as_list(test_string)
+      expect(token_list.map { |token| Tokens.problem?(token) }).to include(true)
+    end
+  end
+
+  ####################
+  # Whitespace
+  ####################
+  context "tokenize empty string" do
+    let(:test_string) { "" }
+    let(:expected_tokens) { [] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize newlines" do
+    let(:test_string) { "\n\n" }
+    let(:expected_tokens) { [TestUtils.token_line(1),
+                             TestUtils.token_line(2)] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize unquoted text should trim spaces" do
+    let(:test_string) { "    foo     \n" }
+    let(:expected_tokens) { [TestUtils.token_unquoted("foo"), TestUtils.token_line(1)] }
+
+    include_examples "token_matching"
+  end
+
+  context "tokenize unquoted text should trim spaces, keep internal spaces" do
+    let(:test_string) { "    foo  bar baz   \n" }
+    let(:expected_tokens) { [TestUtils.token_unquoted("foo"),
+                             TestUtils.token_unquoted(" "),
+                             TestUtils.token_unquoted("bar"),
+                             TestUtils.token_unquoted(" "),
+                             TestUtils.token_unquoted("baz"),
+                             TestUtils.token_line(1)] }
+
+    include_examples "token_matching"
+  end
+
+  ####################
+  # Booleans
```

Comment:
Might want to change this to `Booleans and Null`. This is a nitpick though, so feel free to ignore it